psyneulink/core/llvm/codegen.py

```diff
@@ -21,6 +21,453 @@
+            output_scalar = builder.alloca(ctx.float_ty)
+            # Get the pointer to the first element of the array to convert from [? x double]* -> double*
+            vec_u = builder.gep(x, [ctx.int32_ty(0), ctx.int32_ty(0)])
+            builder.call(ctx.import_llvm_function("__pnl_builtin_vec_sum"), [vec_u, ctx.int32_ty(dim), output_scalar])
+            return output_scalar
+
+        def _tanh(x):
+            output_ptr = builder.alloca(x.type.pointee)
+            helpers.call_elementwise_operation(self.ctx, self.builder, x, helpers.tanh, output_ptr)
+            return output_ptr
+
+        def _exp(x):
+            output_ptr = builder.alloca(x.type.pointee)
+            helpers.call_elementwise_operation(self.ctx, self.builder, x, helpers.exp, output_ptr)
+            return output_ptr
+
+        self.register['sum'] = _vec_sum
+
+        # setup numpy
+        numpy_handlers = {
+            'tanh': _tanh,
+            'exp': _exp
+        }
+
+        for k, v in func_globals.items():
+            if v is np:
+                self.register[k] = numpy_handlers
+
+        name_constants = {
+            True: ir.IntType(1)(1),
+            False: ir.IntType(1)(0),
+        }
+        self.name_constants = name_constants
+        super().__init__()
+
+    def visit_arguments(self, node):
+        args = node.args
+        variable = args[0]
+        # update register
+        self.register[variable.arg] = self.arg_in
+        parameters = args[1:]
+        for param in parameters:
+            assert param.arg not in ["self", "owner"], f"Unable to reference {param.arg} in a compiled UserDefinedFunction!"
+            if param.arg == 'params':
+                assert False, "Runtime parameters are not supported in compiled mode"
+            elif param.arg == 'context':
+                # Since contexts are implicit in the structs in compiled mode, we do not compile it.
+                pass
+            else:
+                self.register[param.arg] = self.func_params[param.arg]
+
+    def _get_type(self, t):
+        type_t = getattr(t, "type", t)
+        if isinstance(type_t, ir.DoubleType):
+            return ArgumentType.SCALAR
+        elif isinstance(type_t, ir.IntType) and type_t.width == 1:
+            return ArgumentType.BOOLEAN
+        elif isinstance(type_t, ir.PointerType):
+            pointee = t.type.pointee
+            return self._get_type(pointee)
+        elif isinstance(type_t, ir.ArrayType):
+            if isinstance(type_t.element, ir.ArrayType):
+                return ArgumentType.MATRIX
+            return ArgumentType.VECTOR
+        else:
+            assert False, "Don't know how to convert LLVM type: {}".format(t)
+
+    def _generate_binop(self, x, y, handlers:dict):
+        # parse args
+        x_type = self._get_type(x)
+        y_type = self._get_type(y)
+        # unpack scalars from pointers
+        if x_type == ArgumentType.SCALAR and type(getattr(x, "type", x)) is ir.PointerType:
```

Comment:
`type(...) is ir.PointerType` should be moved to a helper function and use `isinstance` instead of comparing types.